pkg/storage/mysql/mysqlstore.go

```diff
@@ -0,0 +1,271 @@
+	if err != nil {
+		return nil, fmt.Errorf("failed to use db: %w", err)
+	}
+
+	store := &sqlDBStore{
+		db:     p.db,
+		dbName: name}
+
+	p.dbs[name] = store
+
+	return store, nil
+}
+
+// Close closes the provider.
+func (p *Provider) Close() error {
+	p.Lock()
+	defer p.Unlock()
+
+	for _, store := range p.dbs {
+		err := store.db.Close()
+		if err != nil {
+			return fmt.Errorf(failToCloseProviderErrMsg+": %w", err)
+		}
+	}
+
+	if err := p.db.Close(); err != nil {
+		return err
+	}
+
+	p.dbs = make(map[string]*sqlDBStore)
+
+	return nil
+}
+
+// CloseStore closes a previously opened store
+func (p *Provider) CloseStore(name string) error {
+	p.Lock()
+	defer p.Unlock()
+
+	store, exists := p.dbs[name]
+	if !exists {
+		return nil
+	}
+
+	delete(p.dbs, name)
+
+	conn, err := store.db.Conn(context.Background())
+	if err != nil {
+		return err
+	}
+
+	return conn.Close()
+}
+
+// Put stores the key and the record
+func (s *sqlDBStore) Put(k string, v []byte) error {
+	if k == "" || v == nil {
+		return errors.New("key and value are mandatory")
+	}
+
+	createTableStmt := "CREATE Table IF NOT EXISTS `tablePrefix s.dbName` (`key` varchar(50) NOT NULL, `value` BLOB, " +
+		"PRIMARY KEY (`key`));"
+
+	// creating key-value table inside the database
+	stmt, err := s.db.Prepare(createTableStmt)
+
+	if err != nil {
+		return fmt.Errorf("failed to create table: %w", err)
+	}
+
+	// executing the prepared create table statement
+	_, err = stmt.Exec()
+	if err != nil {
+		return fmt.Errorf("failed to execute the create table statement: %w", err)
+	}
+
+	// prepare upsert query to insert the record, checking whether the key is already mapped to a value in the store.
+	queryStmt, err := s.db.Prepare("INSERT INTO `tablePrefix s.dbName` VALUES (?, ?) ON DUPLICATE KEY " +
+		"UPDATE value=? ")
+	if err != nil {
+		return fmt.Errorf("failed to insert the record in the table %s: %w", `tablePrefix s.dbName`, err)
+	}
+
+	// executing the prepared upsert statement
+	_, err = queryStmt.Exec(k, v, v)
+	if err != nil {
+		return fmt.Errorf("failed to execute the insert query statement %w", err)
+	}
+
+	return queryStmt.Close()
+}
+
+// Get fetches the record based on key
+func (s *sqlDBStore) Get(k string) ([]byte, error) {
+	if k == "" {
+		return nil, errors.New("key is mandatory")
+	}
+
+	var res result
+
+	err := s.db.QueryRow("SELECT * FROM `tablePrefix s.dbName` WHERE `key`=?", k).Scan(&res.key, &res.value)
+	if err != nil {
+		if strings.Contains(err.Error(), sqlDBNotFound) {
+			return nil, storage.ErrDataNotFound
+		}
+
+		return nil, err
+	}
+
+	return res.value, nil
+}
+
+// Delete will delete record with k key
+func (s *sqlDBStore) Delete(k string) error {
+	if k == "" {
+		return errors.New("key is mandatory")
+	}
+
+	res, err := s.db.Exec("DELETE FROM `tablePrefix s.dbName` WHERE `key`=?", k)
+	if err == nil {
+		count, e := res.RowsAffected()
+		if e == nil {
+			if count == 0 {
+				return errors.New("failed to delete row")
```

Comment:
neither the leveldb implementation nor the mem implementation of `Store.Delete()` return an error if the key doesn't exist. I'm not sure if the couchDB one does.

Reply:
Confirmed couch db does provide this feature

Reply:
As discussed this is removed

Reply:
note: #1946 was opened to address this inconsistency